### PR TITLE
Add `warmup` command to build cache explicitly

### LINKED ISSUE #374: Add subcommand to warm-up cache
**What would you like to be added**:
Add subcommand to warm-up cache.

**Why is this needed**:
User may need warm-up cache manually.

### PR DESCRIPTION
**Pull request by Sandy4999**
close #374

### CODE REVIEW

File: pkg/meta/interface.go [hunk 1]
```diff
@@ -209,6 +211,8 @@ type Meta interface {
 	// Setlk sets a file range lock on given file.
 	Setlk(ctx Context, inode Ino, owner uint64, block bool, ltype uint32, start, end uint64, pid uint32) syscall.Errno
 
+	// ListFileSlices returns all slices used by the given file.
+	ListFileSlices(ctx Context, inode Ino, size uint64, slices *[]Slice) syscall.Errno
```
Comment (davies) [inline pkg/meta/interface.go:215]: Once we know the length of a file, we should get all the slices using m.Read()

File: pkg/chunk/cached_store.go [hunk 2]
```diff
@@ -802,4 +802,16 @@ func (store *cachedStore) Remove(chunkid uint64, length int) error {
...
 }
 
+func (store *cachedStore) FillCache(chunkid uint64, length uint32) error {
+	if length == 0 {
+		return nil
+	}
+	r := chunkForRead(chunkid, int(length), store)
+	lastIndx := int(length-1) / store.conf.BlockSize
+	for i := 0; i <= lastIndx; i++ {
+		store.fetcher.fetch(r.key(i))
```
Comment (davies) [inline pkg/chunk/cached_store.go:812]: The fetcher will drop the key if queue is full

Comment (Sandy4999) [reply pkg/chunk/cached_store.go]: 397d6e2

File: pkg/meta/interface.go (hunk 1 above)
Comment (Sandy4999) [reply pkg/meta/interface.go]: c4ddadc

File: cmd/warmup.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		ArgsUsage: "[PATH ...]",
+		Action:    warmup,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "listfile",
+				Aliases: []string{"f"},
+				Usage:   "file containing a list of paths",
+			},
+			&cli.UintFlag{
+				Name:    "concurrent",
```
Comment (davies) [inline cmd/warmup.go:155]: -p --threads

File: cmd/warmup.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+func warmupFlags() *cli.Command {
+	return &cli.Command{
+		Name:      "warmup",
+		Usage:     "build cache for target directories/files",
+		ArgsUsage: "[PATH ...]",
+		Action:    warmup,
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:    "listfile",
```
Comment (davies) [inline cmd/warmup.go:150]: -f --file

File: pkg/chunk/cached_store.go [hunk 5]
```diff
@@ -704,7 +716,7 @@ func NewCachedStore(storage object.ObjectStorage, config Config) ChunkStore {
 		}
 		p := NewOffPage(size)
 		defer p.Release()
-		_ = store.load(key, p, true)
+		_ = store.load(key, p, true, false)
```
Comment (davies) [inline pkg/chunk/cached_store.go:719]: true, don't drop fetched blocks

File: pkg/vfs/fill.go [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+	var attr = &Attr{}
+	for _, p := range paths {
+		if st := resolve(p, &inode, attr); st != 0 {
+			logger.Warnf("Failed to resolve path %s: %s", p, st)
+			continue
+		}
+		logger.Debugf("Warming up path %s", p)
+		if attr.Typ == meta.TypeDirectory {
+			walkDir(inode, todo)
+		} else if attr.Typ != meta.TypeSymlink {
```
Comment (davies) [inline pkg/vfs/fill.go:66]: attr.Type == meta.TypeFile

**Review by davies — APPROVED** [review]

**Pull request merged**